motionplan/armplanning/plan_test.go

```diff
@@ -133,3 +139,259 @@ func TestPlanStep(t *testing.T) {
 		}
 	})
 }
+
+// BenchmarkDanGoalMetric
+//
+// Old:
+// - BenchmarkDanGoalMetric-16                 	  532743	      3448 ns/op	    2416 B/op	      42 allocs/op
+// New:
+// - BenchmarkDanGoalMetric-16                 	  801975	      1247 ns/op	     800 B/op	      14 allocs/op
+// - BenchmarkDanGoalMetric-16                 	  940275	      1145 ns/op	     640 B/op	      11 allocs/op.
+func BenchmarkDanGoalMetric(b *testing.B) {
+	ctx := b.Context()
+	_ = ctx
+	goalFrame := referenceframe.NewPoseInFrame(
+		"world",
+		&spatialmath.DualQuaternion{
+			Number: dualquat.Number{
+				Real: quat.Number{
+					Real: 0.0051161120465661614, Imag: 0, Jmag: 0.9999869126131237, Kmag: 0,
+				},
+				Dual: quat.Number{
+					Real: -798.2045339836992, Imag: 241.2609122119685, Jmag: 4.083757277649134, Kmag: -822.820958100296,
+				},
+			},
+		},
+	)
+	goalFrame.SetName("xarm6")
+
+	options := &PlannerOptions{
+		GoalMetricType: motionplan.SquaredNormOpt,
+	}
+
+	armModel, err := referenceframe.ParseModelJSONFile(utils.ResolveFile("components/arm/fake/kinematics/xarm6.json"), "xarm6")
+	test.That(b, err, test.ShouldBeNil)
+
+	// Create a temporary frame system for the transformation
+	fs := referenceframe.NewEmptyFrameSystem("")
+	err = fs.AddFrame(armModel, fs.World())
+	test.That(b, err, test.ShouldBeNil)
+
+	// metricFn := options.getGoalMetric(referenceframe.FrameSystemPoses{"xarm6": goalFrame})
+	// inps := referenceframe.FrameSystemInputs{"xarm6": []referenceframe.Input{
+	//  	-1.335, -1.334, -1.339, -1.338, -1.337, -1.336,
+	// }}
+	// ans := metricFn(&motionplan.StateFS{
+	//  	Configuration: inps,
+	//  	FS:            fs,
+	// })
+	// test.That(b, ans, test.ShouldEqual, 6.1075976675485745e+06)
+	//
+	// for b.Loop() {
+	//  	metricFn(&motionplan.StateFS{
+	//  		Configuration: inps,
+	//  		FS:            fs,
+	//  	})
+	// }
+
+	metricFn, err := options.getGoalMetricLinear(referenceframe.FrameSystemPoses{"xarm6": goalFrame})
+	test.That(b, err, test.ShouldBeNil)
+
+	inps := []referenceframe.Input{
+		-1.335, -1.334, -1.339, -1.338, -1.337, -1.336,
+	}
+	ans := metricFn(&motionplan.LinearFS{
+		Configuration: inps,
+		FS:            fs,
+	})
+	test.That(b, ans, test.ShouldEqual, 6.1075976675485745e+06)
```

Comment:
Is this deterministic enough to pass a float equality check? Should we not be using `almostequals`?

Reply:
Funny you didn't ask how I knew what this number should be! It's the actual magic number derived from running the old code to make sure my changes computed the right answers.

Almost equals is probably right here.